gp-inventory/gpi-waiting-list.php

```diff
@@ -8,99 +8,197 @@
  * Replace the default available inventory message with a waiting list message when a choice's inventory is exhausted.
  * The choice will be selectable and submittable.
  */
-add_filter( 'gpi_disable_choices', '__return_false' );
-add_filter( 'gpi_remove_choices', '__return_false' );
-
-function gpiw_apply_waitlist_message( $choice, $field, $form, $how_many_left = false ) {
-	$message         = '(waiting list)';
-	$default_message = gp_inventory_type_choices()->replace_choice_available_inventory_merge_tags( gp_inventory_type_choices()->get_inventory_available_message( $field ), $field, $form, $choice, $how_many_left );
-	if ( strpos( $choice['text'], $default_message ) === false ) {
-		$choice['text'] .= ' ' . $message;
-	} else {
-		$choice['text'] = str_replace( $default_message, $message, $choice['text'] );
+class GPI_Waiting_List {
+	public $waitlist_message = '(waiting list)';
+
+	public function __construct() {
+		add_action( 'init', array( $this, 'add_hooks' ), 20 ); // Wait for GPI to be instantiated
 	}
-	return $choice;
-}
 
-add_filter( 'gpi_pre_render_choice', function( $choice, $exceeded_limit, $field, $form, $count ) {
+	public function add_hooks() {
+		add_filter( 'gform_entry_post_save', array( $this, 'add_entry_meta' ), 10, 2 );
+
+		/*
+		 * Choice-based hooks
+		 */
+		add_filter( 'gpi_disable_choices', '__return_false' );
+		add_filter( 'gpi_remove_choices', '__return_false' );
+
+		add_filter( 'gpi_pre_render_choice', array( $this, 'pre_render_choice' ), 10, 5 );
+
+		// Add support for showing waiting list message in confirmations and notifications.
+		add_filter( 'gform_pre_submission_filter', array(
+			$this,
+			'add_waitlist_message_to_choices_on_submission'
+		), 10, 2 );
+
+		add_filter( 'gform_entries_field_value', function ( $value, $form_id, $field_id, $entry ) {
```

Comment:
Not a deal breaker but it does feel a little strange to be using anonymous functions now that this is a class?